// app/server/routes/webapp/index.js
var webapp_routes = module.exports = require('express').Router(),
    
    path_to_client_dir = __dirname + '/../../../client';

// parse form submissions
  webapp_routes.use( require('body-parser').urlencoded({ extended: false }) );

// handle sessions
  webapp_routes.use( require('express-session')({

    secret: process.env.WEBAPP_SESSION_SECRET, 
    resave: false, 
    saveUninitialized: false, 
    cookie: { maxAge: 1000 * 60 * 5 }}) 
  );

// routes

  // main
    webapp_routes.get('/', function( req, res ){

      var React = require('react'),
          ReactDOMServer = require('react-dom/server'),

          home_layout = React.createElement( require('app/client/src/js/home-page'), {

            credentials_error: req.query.hasOwnProperty('bad-credentials'),
            server_error: req.query.hasOwnProperty('server-error')
          });

      res.render( path_to_client_dir + '/home.ejs', { react_server_render: ReactDOMServer.renderToString( home_layout ) });
    });

  // login
    webapp_routes.post('/login', function( req, res ){
    
      var readability = require('app/server/modules/readability');

      readability.login( req.body.user, req.body.pass, function( error, tokens ){

        if( error ){

          var error_string = error.toString();

          if( error_string.indexOf('HTTP 401') > -1 ) return res.redirect('/?bad-credentials');

          else return res.redirect('/?server-error');
        }

        req.session.auth = tokens;

        res.redirect('/bookmarks');
      });
    });

  // list bookmarks
    webapp_routes.get('/bookmarks/:page?', function( req, res ){

      if( !req.session.auth ) return res.redirect('/');
    
      var readability = require('app/server/modules/readability'),
          page = req.params.page || 1;

      readability.bookmarks( req.session.auth.oauth_token, req.session.auth.oauth_token_secret, { page: page }, function( error, response ){

        var React = require('react'),
            ReactDOMServer = require('react-dom/server'),
            index_layout = React.createElement( require('app/client/src/js/index-page'), {

              bookmarks: response.bookmarks,
              metadata: response.meta
            });       

        res.set('Content-Type', 'text/html; charset=utf-8');

        res.render( path_to_client_dir + '/index.ejs', { react_server_render: ReactDOMServer.renderToString( index_layout ) });
      });
    });

  // get article
    webapp_routes.get('/article/:id', function( req, res ){

      if( !req.session.auth ) return res.redirect('/');
    
      var readability = require('app/server/modules/readability');

      readability.article( req.session.auth.oauth_token, req.session.auth.oauth_token_secret, req.params.id, function( error, article ){

        if( error || !article ) return res.redirect( 'back' );

        var React = require('react'),
            ReactDOMServer = require('react-dom/server'),
            article_layout = React.createElement( require('app/client/src/js/article-template'), {

              source_url: article.url,
              source_domain: article.domain,
              title: article.title,
              content: article.content
            });       

        res.render( path_to_client_dir + '/article.ejs', { 
          
          article_title: article.title,
          react_server_render: ReactDOMServer.renderToString( article_layout )
        });
      });
    });
