// app/server/modules/readability/bookmark-actions.js
module.exports = function( readability ){
  
  return {

    add: function( token, secret, url, callback ){

      if( !callback || typeof callback !== 'function' ){ callback = function(){}; }

      var reader = new readability.reader({

        access_token: token,
        access_token_secret: secret
      });

      reader.addBookmark( url, function( error, bookmark ){

        callback( error, bookmark );
      });
    },

    remove: function( token, secret, id, callback ){

      if( !callback || typeof callback !== 'function' ){ callback = function(){}; }

      var reader = new readability.reader({

        access_token: token,
        access_token_secret: secret
      });

      reader.removeBookmark( id, function( error, is_removed ){

        callback( error, is_removed );
      });
    },

    favorite: function( token, secret, id, callback ){

      if( !callback || typeof callback !== 'function' ){ callback = function(){}; }

      var reader = new readability.reader({

        access_token: token,
        access_token_secret: secret
      });

      reader.favoriteBookmark( id, function( error, bookmark ){

        callback( error, bookmark );
      });
    },

    unfavorite: function( token, secret, id, callback ){

      if( !callback || typeof callback !== 'function' ){ callback = function(){}; }

      var reader = new readability.reader({

        access_token: token,
        access_token_secret: secret
      });

      reader.unfavoriteBookmark( id, function( error, bookmark ){

        callback( error, bookmark );
      });
    },

    archive: function( token, secret, id, callback ){

      if( !callback || typeof callback !== 'function' ){ callback = function(){}; }

      var reader = new readability.reader({

        access_token: token,
        access_token_secret: secret
      });

      reader.archiveBookmark( id, function( error, bookmark ){

        callback( error, bookmark );
      });
    },

    unarchive: function( token, secret, id, callback ){

      if( !callback || typeof callback !== 'function' ){ callback = function(){}; }

      var reader = new readability.reader({

        access_token: token,
        access_token_secret: secret
      });

      reader.unarchiveBookmark( id, function( error, bookmark ){

        callback( error, bookmark );
      });
    }
  }
}
