// app/server/modules/readability/index.js
var readability = require('app/server/modules/patched-readability-api');

readability.configure({
  consumer_key: process.env.READABILITY_READER_API_KEY,
  consumer_secret: process.env.READABILITY_READER_API_SECRET,
  parser_token: process.env.READABILITY_PARSER_API_KEY
});

module.exports = {
  
  login: function login( username, password, callback ){

    readability.xauth( username, password, function( error, tokens ){

      if( error ) return callback( error );
      
      callback( null, tokens );
    });
  },

  bookmarks: function get_bookmarks( token, secret, params, callback ){

    var reader = new readability.reader({

      access_token: token,
      access_token_secret: secret
    });

    reader.bookmarks( params, function( error, bookmarks ){
      
      if( error ) return callback( error );

      callback( null, bookmarks );
    });
  },

  article: function get_article( token, secret, id, callback ){

    var reader = new readability.reader({

      access_token: token,
      access_token_secret: secret
    });

    reader.article( id, function( error, article ){

      if( error ) return callback( error );

      callback( null, article );
    });
  },

  bookmark: require('app/server/modules/readability/bookmark-actions')( readability )
}
