// app/client/src/js/events/view/index.js
module.exports = function( noticeboard ){

  // set current view
    var path =  window.location.pathname;

    if( path === '/' ) noticeboard.notify('update-app-state', { view: 'home' }, 'view:bootstrap');

    else if( path.indexOf('/bookmarks') === 0 ) noticeboard.notify('update-app-state', { view: 'index' }, 'view:bootstrap');
    
    else if( path.indexOf('/favorites') === 0 ) noticeboard.notify('update-app-state', { view: 'index', view_data: { filter: 'favorites' } }, 'view:bootstrap');

    else if( path.indexOf('/archives') === 0 ) noticeboard.notify('update-app-state', { view: 'index', view_data: { filter: 'archives' } }, 'view:bootstrap');

    else if( path.indexOf('/article/') === 0 ){

      var article_id = path.replace('/article/', '');

      noticeboard.notify('update-app-state', { view: 'article', view_data: { id: article_id } }, 'view:bootstrap');
    }
  
  noticeboard.once('bootstrap-completed', 'default-view-behavior-after-bootstrap', function( msg ){

    noticeboard.notify('get-app-state', function( state ){

      if( state.last_view ) return;

      switch( state.view ){

        case 'home':
          
          noticeboard.notify('next-view', { view: 'index' }, 'view:default-view-behavior-after-bootstrap');

        break;
      }
    });
  });

  noticeboard.watch('next-view', 'change-to-next-view', function( msg ){

    var next_view = msg.notice.view,
        next_view_data = msg.notice.view_data || {};

    noticeboard.notify('get-app-state', function( state ){

      var current_view = state.view,
          current_view_data = state.view_data;

      switch( next_view ){

        case 'index':

          var get_bookmarks = {};
              get_bookmarks.type = next_view_data.filter || 'bookmarks';
              get_bookmarks.page = next_view_data.page || 1;
              get_bookmarks.callback = function(){

                var has_executed = false;

                return function( data ){

                  if( has_executed ) return;

                  var React = require('react'),
                      ReactDOM = require('react-dom'),

                      index_layout = React.createElement( require('app/client/src/js/index-page'), {

                        bookmarks: data.bookmarks,
                        metadata: data.meta,
                        page: get_bookmarks.type
                      });

                  noticeboard.once('teardown-view', 'render-next-view', function(){

                    ReactDOM.render( index_layout, document.getElementById('main-wrapper'), function(){

                      noticeboard.notify('update-app-state', {

                        last_view: current_view, 
                        last_view_data: current_view_data,

                        current_view: next_view,
                        current_view_data: next_view_data

                      }, 'view:change-to-next-view');

                      noticeboard.watch( 'bookmarks-pages-updated', 'sync-layout-with-pages', function(){

                        noticeboard.notify('get-bookmarks-page', {

                          type: get_bookmarks.type,
                          page: get_bookmarks.page,
                          callback: function( data ){

                            var React = require('react'),
                                ReactDOM = require('react-dom'),

                                index_layout = React.createElement( require('app/client/src/js/index-page'), {

                                  bookmarks: data.bookmarks,
                                  metadata: data.meta,
                                  page: get_bookmarks.type
                                });
                            
                            ReactDOM.render( index_layout, document.getElementById('main-wrapper') );
                          }
                        });
                      });

                      noticeboard.once('teardown-view', 'index-view-teardown', function(){

                        noticeboard.ignore('bookmarks-pages-updated', 'sync-layout-with-pages');
                      });
                    });
                  });
                  
                  noticeboard.notify('teardown-view', null, 'change-to-next-view');
                }
              }();

          noticeboard.notify('get-bookmarks-page', get_bookmarks);

        break;
      }
    }, 'view:change-to-next-view');

  })
}
